### PR TITLE
WIP Investiagte dqn

### PR DESCRIPTION
**Pull request by mesjou**

### CODE REVIEW

**Comment by mesjou** [comment]
@SimonHashtag mostly refactoring. Real changes
- ReLu activation function instead of Tanh? don't know if important with only 2 hidden layers but possible pitfall if we use really deep networks in the future (vanishing gradients)
- added DDQN option

File: lifecycle/train_dqn.py [hunk 1]
```diff
@@ -1,78 +1,89 @@
...
+            loss = agent.learn(obs, next_obs, action, reward, done)
+            obs = next_obs  # update observations
+
+            # logging
+            rewards.append(reward)
+            losses.append(loss if isinstance(loss, int) else loss.numpy())
+            iterator += 1
+
+            # update target network if it was initialized already
+            if iterator % copy_step == 0 and len(agent.target_network.get_weights()) > 0:
```
Comment (SimonHashtag) [inline lifecycle/train_dqn.py:58]: How do we make sure the target_network is initialized?

File: lifecycle/agents/dqn.py [hunk 2]
```diff
@@ -1,124 +1,141 @@
 import numpy as np
 import tensorflow as tf
 import random
+from typing import Dict
+from tensorflow.keras.initializers import RandomNormal
+
+from lifecycle.utils.buffer import ReplayBuffer
+from lifecycle.utils.losses import huber_loss
```
Comment (SimonHashtag) [inline lifecycle/agents/dqn.py:8]: Let's implement huber loss in loss function of dqn before merging, @mesjou? (as argument in dqn class)

**Review by SimonHashtag — COMMENTED** [review]
Two issues:

1. Possible implementation of huber-loss before merge?
2. Question on how we initialize target network.

File: lifecycle/train_dqn.py (hunk 1 above)
Comment (mesjou) [reply lifecycle/train_dqn.py:58]: if it is not initialised get_weight() returns empty list -> weight get only update if one forward pass through target network was performed and, thus, len(get_weight()) > 0. This is only the case when agent learned at least one round. before this time network weights of the local network stay the same anyways and we do not need to update

File: lifecycle/agents/dqn.py (hunk 2 above)
Comment (mesjou) [reply lifecycle/agents/dqn.py:8]: we can do that but in my experiments the performance did not really increase. I am not sure if we will need it in the future

Comment (SimonHashtag) [reply lifecycle/agents/dqn.py:8]: Alright. Then, let's do it later

File: lifecycle/train_dqn.py (hunk 1 above)
Comment (mesjou) [reply lifecycle/train_dqn.py:58]: @SimonHashtag do you have a more explicit idea? it is not super explicit but it will work

Comment (SimonHashtag) [reply lifecycle/train_dqn.py:58]: Ah, now I get it. I didn't understand how the target_network gets any weights. But of course this happens, whenever it calculates the value for the next state (value_next)...

**Review by SimonHashtag — APPROVED** [review]
Got DQN working! 
- **fixed action mask**
- Added replay buffer
- Added option for double dqn
- Included function life in run_dqn
- Added utils dir with replay buffer, loss function file, plotting file, seeding file
